src/s390x/zipl.rs

```diff
@@ -29,66 +38,232 @@ pub fn chreipl<P: AsRef<Path>>(dev: P) -> Result<()> {
+    }
+}
+
+fn find_files<P: AsRef<Path>>(
+    path: P,
+    f: impl Fn(&DirEntry) -> Option<PathBuf>,
+) -> Result<Vec<PathBuf>> {
+    Ok(read_dir(&path)
+        .with_context(|| format!("reading directory {}", path.as_ref().display()))?
+        .collect::<std::io::Result<Vec<_>>>()?
+        .iter()
+        .filter_map(f)
+        .collect::<Vec<_>>())
+}
+
+fn regenerate_initrd<P: AsRef<Path>>(source: P, files: &[PathBuf]) -> Result<NamedTempFile> {
+    let source = source.as_ref();
+    let dest = Builder::new()
+        .prefix("initrd")
+        .suffix(".img")
+        .append(true)
+        .tempfile()
+        .context("creating cpio for extras")?;
+
+    // copying from original initrd, on s390x SE it's not concatenated
+    std::fs::copy(source, dest.path())
+        .with_context(|| format!("copying {} to {}", source.display(), dest.path().display()))?;
+
+    // new small initrd with files
+    let mut ino = 0;
+    let mut initrd = GzEncoder::new(dest.as_file(), Compression::default());
+    for path in files {
+        ino += 1;
+        let file = File::open(path)?;
+        let meta = file.metadata()?;
+        let name = &path.to_string_lossy();
+        let mode = meta.permissions().mode();
+        let mut input = BufReader::new(file);
+        let entry = NewcBuilder::new(name)
+            .mode(mode as u32)
+            .uid(meta.uid())
+            .gid(meta.gid())
+            .mtime(meta.mtime() as u32)
+            .ino(ino);
+        let size = if mode & 0o40_000 == 0o40_000 {
+            // directory
+            0
+        } else {
+            meta.len()
+        };
+        let mut writer = entry.write(&mut initrd, size as u32);
+        let mut tmp = [0u8; 1024];
+        copy_exactly_n(&mut input, &mut writer, size as u64, &mut tmp)?;
+        writer.finish()?;
+    }
+    // finishing CPIO
+    trailer(&mut initrd)?;
+    initrd.finish()?.flush()?;
+
+    Ok(dest)
+}
+
+fn get_info_from_bls(boot: &Path) -> Result<(String, String, String)> {
+    let mut kernel = None;
+    let mut initrd = None;
+    let mut options = None;
+
+    let read_opts = |contents: &str| {
+        for l in contents.lines() {
+            match l.split_once(' ') {
+                Some(("linux", s)) => kernel = Some(s.trim().to_owned()),
+                Some(("initrd", s)) => initrd = Some(s.trim().to_owned()),
+                Some(("options", s)) => options = Some(s.trim().to_owned()),
+                _ => {}
+            }
+        }
+        Ok(None)
+    };
+    visit_bls_entry(boot, read_opts)?;
+
+    let kernel = kernel.ok_or_else(|| anyhow!("missing 'linux' key in default BLS config"))?;
+    let initrd = initrd.ok_or_else(|| anyhow!("missing 'initrd' key in default BLS config"))?;
+    let options = options.ok_or_else(|| anyhow!("missing 'options' key in default BLS config"))?;
+
+    let kernel = format!("{}{}", boot.to_string_lossy(), kernel);
+    let initrd = format!("{}{}", boot.to_string_lossy(), initrd);
```

Comment:
Let's avoid using `to_string_lossy`. Those paths should really never be non-UTF8 but it's easy to do the right thing: we can do `boot.join(kernel)` and `boot.join(initrd)` and return `PathBuf`s. I think all the spots where `kernel` and `initrd` are used in the caller support `Path`s already.

Reply:
> we can do boot.join(kernel)

actually not:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2e566c75273b264d50927b6102ef2644

we'd have to remove `/` , so imo it's easier to use `format!`